protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/ReceivedDataLineFilter.java

```diff
@@ -120,12 +125,102 @@ protected Location getLocation() {
         header.add(headerLineBuffer.toString());
         headerLineBuffer = new StringBuilder();
 
-        headerLineBuffer.append(DATEFORMAT.get().format(new Date()));
+        headerLineBuffer.append(DATEFORMAT.format(ZonedDateTime.now()));
 
         header.add(headerLineBuffer.toString());
         
-        return Arrays.asList(header);
+        return Collections.singletonList(header);
     
     }
-    
+
+    @Override
+    protected Response onSeparatorLine(SMTPSession session, ByteBuffer line, LineHandler<SMTPSession> next) {
+        if (getLocation() == Location.Suffix && !session.getAttachment(headersSuffixAdded, State.Transaction).isPresent()) {
+            session.setAttachment(headersSuffixAdded, Boolean.TRUE, State.Transaction);
+            return addHeaders(session, line, next);
+        }
+        return super.onSeparatorLine(session, line, next);
+    }
+
+    @Override
+    protected Response onHeadersLine(SMTPSession session, ByteBuffer line, LineHandler<SMTPSession> next) {
+        if (getLocation() == Location.Prefix && !session.getAttachment(headersPrefixAdded, State.Transaction).isPresent()) {
+            session.setAttachment(headersPrefixAdded, Boolean.TRUE, State.Transaction);
+            return addHeaders(session, line, next);
+        }
+        return super.onHeadersLine(session, line, next);
+    }
+
+    /**
+     * Add headers to the message
+     *
+     * @return response
+     */
+    private Response addHeaders(SMTPSession session, ByteBuffer line, LineHandler<SMTPSession> next) {
+        Response response;
+        for (Header header: headers(session)) {

```

Comment:
lacking space before `:`

Reply:
We currently have both forms (and I didn't notice it for now)
Let's not change it for now and change Checkstyle if we think it's important.